<a href="https://colab.research.google.com/github/smraytech/test/blob/main/BankChurnersProjectMLandAI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem statement



https://www.kaggle.com/datasets/sakshigoyal7/credit-card-customers
Predict Churning customers
This Dataset has attrited and existing customers.Bank management would like to know the attrited customers such that it can improve on service delivery.
Analyse & get get out insights to help management improve on operational efficiciency & take vital dicisions hence retaining the attrited customers and realising profitibility.

In [ ]:
import pandas as pd
import numpy as np
import sklearn.model_selection
import sklearn.compose
import sklearn.preprocessing
import sklearn.svm
from sklearn.svm import SVC
import sklearn.tree
import sklearn.neighbors
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score


*  I uploaded the BankChurners Dataset
*   To make sure that all the columns are vissible i used this code
pd.set_option('display.max_columns', None)



In [ ]:
pd.set_option('display.max_columns', None)
df_f = pd.read_csv("/content/BankChurners.csv")


# Data Exploration

This method displays the first five columns,starting from index 0 to index 4

In [ ]:
df_f.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  Total_Relationship_Count  Months_Inactive_12_mon  \
0              39                         5                       1   
1              44                         6                       1   
2              36                         4                       1   
3              34                         3                       4   
4              21                         5                       1   

   Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0                      3       12691.0                  777          11914.0   
1                      2        8256.0                  864           7392.0   
2                      0        3418.0                    0           3418.0   
3                      1        3313.0                 2517            796.0   
4                      0        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998

Displaying the summary of either null or not null values

In [ ]:
df_f.isnull().sum()

CLIENTNUM                                                                                                                             0
Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                 

Displaying the different Data types,we can see that our dataset has int64,float64, and object.

In [ ]:
df_f.dtypes

CLIENTNUM                                                                                                                               int64
Attrition_Flag                                                                                                                         object
Customer_Age                                                                                                                            int64
Gender                                                                                                                                 object
Dependent_count                                                                                                                         int64
Education_Level                                                                                                                        object
Marital_Status                                                                                                                         object
Income

Therefore the unique() method returns Existing Customer and Attrited Customer.

In [ ]:
df_f['Attrition_Flag'].unique()

array(['Existing Customer', 'Attrited Customer'], dtype=object)

In [ ]:
df_f[df_f['Attrition_Flag'] == 'Attrited Customer'].head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
21  708508758  Attrited Customer            62      F                0   
39  708300483  Attrited Customer            66      F                0   
51  779471883  Attrited Customer            54      F                1   
54  714374133  Attrited Customer            56      M                2   
61  712030833  Attrited Customer            48      M                2   

   Education_Level Marital_Status Income_Category Card_Category  \
21        Graduate        Married  Less than $40K          Blue   
39       Doctorate        Married         Unknown          Blue   
51        Graduate        Married  Less than $40K          Blue   
54        Graduate        Married         $120K +          Blue   
61        Graduate        Married     $60K - $80K        Silver   

    Months_on_book  Total_Relationship_Count  Months_Inactive_12_mon  \
21              49                         2                       3   
39              56                         5                       4   
51              40                         2                       3   
54              36                         1                       3   
61              35                         2                       4   

    Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
21                      3        1438.3                    0           1438.3   
39                      3        7882.0                  605           7277.0   
51                      1        1438.3                  808            630.3   
54                      3       15769.0                    0          15769.0   
61                      4       34516.0                    0          34516.0   

    Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
21                 1.047              692              16   
39                 1.052              704              16   
51                 0.997              705              19   
54                 1.041              602              15   
61                 0.763              691              15   

    Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  \
21                0.600                  0.000   
39                0.143                  0.077   
51                0.900                  0.562   
54                0.364                  0.000   
61                0.500                  0.000   

    Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
21                                            0.99616                                                                                    
39                                            0.99780                                                                                    
51                                            0.99028                                                                                    
54                                            0.99671                                                                                    
61                                            0.99823                                                                                    

    Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
21                                           0.003836                                                                                   
39                                           0.002197                                                                                   
51                                           0.009720                                                                                   
54                                           0.003294                                                                                   
61                                           0.001771

## Data Cleaning
We do data cleaning because we do not want to carry out predictions with unclean dataset because it will give us wrong predictions hence missleading organisation decisions.

Dropping two columns from our dataset by using df.drop method

In [ ]:
df_f.drop(['CLIENTNUM','Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1','Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'], axis=1,inplace=True)


Total number of columns after dropping two columns fro our dataset by using
 len() function

In [ ]:
len(df_f.columns)

20

Data Preprocessing.
In this process i will do more cleaning,prepare my raw data such that i can carry out feature engineering.

x1_train will contain all the data from my dataframe apart from from the Attrition _Frag column which i have dropped.
Attrition_Frag column data will now be in the y1_train.

In [ ]:
x1_train = df_f.drop(['Attrition_Flag'],axis=1)
y1_train = df_f['Attrition_Flag']
x1_train.shape


(10127, 19)

I split my Dataset into training & test sets,75%  and 25% respectively

In [ ]:
x1_train,x1_test,y1_train,y1_test = sklearn.model_selection.train_test_split(x1_train,y1_train,test_size=0.25)
print('x1_train.shape:',x1_train.shape)
print('x1_test.shape:',x1_test.shape)
print('y1_train.shape:',y1_train.shape)
print('y1_test.shape:',y1_test.shape)

x1_train.shape: (7595, 19)
x1_test.shape: (2532, 19)
y1_train.shape: (7595,)
y1_test.shape: (2532,)


I select columns from my dataset x1_train that is to say type 'int64' & 'float64' by 'select_dtypes' method

I also select columns from dataset x1_train with the object type,majorly composed of categorical data.

sklearn.preprocessing.OneHotEncoding transforms categorical attributes.




In [ ]:
nattributes_attributes = x1_train.select_dtypes(include=['int64','float64']).columns
cattributes_attributes = x1_train.select_dtypes(include='object').columns
ctran = sklearn.compose.ColumnTransformer([
  ('standard_scaling',sklearn.preprocessing.StandardScaler(),nattributes_attributes),
  ('one_hot_encoding',sklearn.preprocessing.OneHotEncoder(handle_unknown='ignore'),['Gender','Education_Level','Marital_Status','Income_Category','Card_Category','Income_Category'])
])


After executing these steps,x1_train and x1_test will be preprocessed according to the transformations specified in the ColumnTransformer hence ready for further use in machine learning tasks.



In [ ]:
ctran.fit(x1_train)
x1_train = ctran.transform(x1_train)
x1_test = ctran.transform(x1_test)

# Model Assessment

Model comparison that is  SVM,KNN,and Decision Trees are different algorithms to be used in this case since each performs differently due to built in behaviours & principles possesed by each.This will help us understand which algorithm performs better than the other.



*  The one which fits the problem best will be chosen.
*  By testing all these algorithms it will help us no the insights across various algorithms.

*  Data capture differ with different algorithms.


* With cross-validation,estimating how well models generalize to unseen data will be put into consideration.














## SVM

In [ ]:

parameters_grid = {
    "kernel":["linear","rbf","poly"],
    "C":[0.005,0.05,0.5,5,15,100]
}
m_svm = sklearn.model_selection.GridSearchCV(sklearn.svm.SVC(),parameters_grid,scoring="accuracy",cv=5,n_jobs=-1)
m_svm.fit(x1_train,y1_train)
print("Accuracy of finest svm classifier = {:.2f}".format(m_svm.best_score_))
print("Finest hyperparameters got={}".format(m_svm.best_params_))


Accuracy of finest svm classifier = 0.94
Finest hyperparameters got={'C': 5, 'kernel': 'rbf'}


## Decision Tree

In [ ]:
parameters_grid = {
    "criterion":["gini","entropy"],
    "max_depth":range(3,25,5),
    "min_samples_split":range(3,25,5)
}
m_tree = sklearn.model_selection.GridSearchCV(sklearn.tree.DecisionTreeClassifier(),parameters_grid,scoring="accuracy",cv=5,n_jobs=-1)
m_tree.fit(x1_train,y1_train)
print("Accuracy of finest decision tree classifier = {:.2f}".format(m_tree.best_score_))
print("Finest hyperparameters got ={}".format(m_tree.best_params_))


Accuracy of finest decision tree classifier = 0.94
Finest hyperparameters got ={'criterion': 'entropy', 'max_depth': 18, 'min_samples_split': 18}


## K Nearest Neighbors (KNN)

In [ ]:
parameters_grid = {
    "n_neighbors":[1,6,11,16,21],
    "metric":["minkowski","euclidean","manhattan"]
}
m_knn = sklearn.model_selection.GridSearchCV(sklearn.neighbors.KNeighborsClassifier(),parameters_grid,scoring="accuracy",cv=5,n_jobs=-1)
m_knn.fit(x1_train,y1_train)
print("Accuracy of finest KNN classifier = {:.2f}".format(m_knn.best_score_))
print("Finest hyperparameters got ={}".format(m_knn.best_params_))


Accuracy of finest KNN classifier = 0.90
Finest hyperparameters got ={'metric': 'minkowski', 'n_neighbors': 16}


# Testing the Best Model

1. m_tree.predict(x1_test) uses the trained Tree model m_tree to predict target variable(y_pre) for features in the test dataset(x1_test)
2. sklearn.metrics.accuracy_score(y1_test,y_pre) to calculate accuracy of predictions made.
3. sklearn.metrics.confusion_matrix(y1_test,y_pre) used to compute confusion matrix,a table that summarizes performance of classification model by comparing actual and predicted classes.
4. sklearn.metrics.precision_recall_fscore_support(y1_test,y_pre) computes precision,recall,F1-score and support for each class,returns 4 arrays containing precision,recall,f1-score,and number of true instances for each class.
5. sklearn.metrics.classification_report(y1_test,y_pre),compiles a text report showing precision,recall,f1-score & support for each class and the macro and weighted averanges of these matrics.





In [ ]:
y_pre = m_tree.predict(x1_test)
final_accurcy = sklearn.metrics.accuracy_score(y1_test,y_pre)
confus_ma = sklearn.metrics.confusion_matrix(y1_test,y_pre)
precision,recall,f1,support = sklearn.metrics.precision_recall_fscore_support(y1_test,y_pre)
classific_repo = sklearn.metrics.classification_report(y1_test,y_pre)
print("Accuracy_final=",final_accurcy)
print("Precision_final=",precision)
print("Recall_final=",recall)
print("f1-Score=",f1)
print("Confusion Matrix:\n",confus_ma)
print("Classification Report:\n",classific_repo)

Accuracy_final= 0.9439178515007899
Precision_final= [0.84972678 0.9598338 ]
Recall_final= [0.78140704 0.9742268 ]
f1-Score= [0.81413613 0.96697674]
Confusion Matrix:
 [[ 311   87]
 [  55 2079]]
Classification Report:
                    precision    recall  f1-score   support

Attrited Customer       0.85      0.78      0.81       398
Existing Customer       0.96      0.97      0.97      2134

         accuracy                           0.94      2532
        macro avg       0.90      0.88      0.89      2532
     weighted avg       0.94      0.94      0.94      2532



# Conclusion

 Decision Tree and SVM  got 94% as accuracy while KNN with 90% hence adding more confidance to my findings, agreater strength.

I chose Decision Tree,my Attrited customers's performance with precision=0.85, recall=0.78, f1-score=0.81
The greater the values the better performance.